core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -225,22 +228,36 @@ object RequestChannel extends Logging {
 
   }
 
-  /** responseAsString should only be defined if request logging is enabled */
-  class Response(val request: Request, val responseSend: Option[Send], val responseAction: ResponseAction,
-                 val responseAsString: Option[String]) {
-    request.responseCompleteTimeNanos = Time.SYSTEM.nanoseconds
-    if (request.apiLocalCompleteTimeNanos == -1L) request.apiLocalCompleteTimeNanos = Time.SYSTEM.nanoseconds
+  abstract class Response(val request: Request) {
+    val nowNs = Time.SYSTEM.nanoseconds
+    request.responseCompleteTimeNanos = nowNs
+    if (request.apiLocalCompleteTimeNanos == -1L)
+      request.apiLocalCompleteTimeNanos = nowNs
 
     def processor: Int = request.processor
 
-    override def toString =
-      s"Response(request=$request, responseSend=$responseSend, responseAction=$responseAction), responseAsString=$responseAsString"
+    def toString(responseSend: Option[Send], responseAsString: Option[String]): String = {
+      s"Response(responseType=${this.getClass} request=$request, responseSend=$responseSend), responseAsString=$responseAsString"
```

Comment:
nit: it's a little annoying that we will print `responseSend=None` for anything not a `SendResponse`. Also using `Class.toString` is probably unnecessarily verbose. I think I preferred how you initially structured it where each subclass had its own implementation. It's not a lot of code we're saving here.